segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/WriterState.java

```diff
@@ -0,0 +1,292 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.server.host.handler;
+
+import com.google.common.base.Preconditions;
+import io.pravega.segmentstore.contracts.BadOffsetException;
+import io.pravega.shared.protocol.netty.WireCommands;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * Maintains state for a Writer in the {@link AppendProcessor}.
+ * This class is tightly-coupled with the {@link AppendProcessor} behavior and should not be used for any other purpose.
```

Comment:
Indeed it is. Perhaps we could separate concerns a little bit: The main complication that is making things difficult is the possibility of threads reordering things. What if we used this class to order the responses of all of the replies successful or not. This could be done similar to how it is here with the list of runnables. If however each were associated with an event number we could make sure they are run in the correct order.
Obviously the two classes would still be fairly closely tied because of locking etc. But it might make it easier to reason about the correctness.

Reply:
I'm answering both this and https://github.com/pravega/pravega/pull/4418#discussion_r354090715.

Your solution will indeed solve the correctness of the general case. However, the downside is the need for an array/list/queue for each `WriterState` where we'll keep all the in-flight append event numbers. Assuming an average array size of 256 entries (array expands, but doesn't contract), we will need 256 x 8 bytes x 10000 connections = 20MB of heap just to store this.

Taking a step back, this problem (with a successful append acking a possible failed one) exists even now in master. This PR doesn't make the problem worse. The problem was introduced with the pipelining.

However let's think practically. In reality we'll only have SegmentSealed and BadAttribute.

If we get SegmentSeal then any append after that will also be failed. None will succeed.
For BadAttribute it's the same story. The client isn't going to send a random bad one followed by a good one. It will send a flurry of bad ones.

The main goal of this PR was to make sure that any error messages are sent back to the client only after all previous appends have been acked, otherwise we risk duplicating them Inna segment's successors. 

If that had been enough this PR would have been simpler. However for some reason I don't fully understand, some integration tests expect that I send all error messages. I did a dump of what exceptions are generated in those tests and it was always SegmentIsSealed. 

So while I agree with your assessment that this method is not correct in the general case, when factoring in the behavior of the client and the Appendprocessor, it will work as expected.

Reply:
We have discussed this offline and agreed that, barring the Segment Store works incorrectly, this should not be an issue.

I've added a `log.warn` to indicate if we get any successful appends after a failed one via a Connection/Writer.